harvester/wacz.py

```diff
@@ -0,0 +1,273 @@
+    HTTP requests made while crawling the pages.  The fully rendered form of the pages
+    are stored in binary WARC files, with "index" files that help correlate a particular
+    URL with the location in those WARC files.  This client supports reading files from
+    a WACZ archive, and some utilities to actually extract the rendered page HTML from
+    the WARC files.
+
+    NOTE: This class is designed to be used via a context manager, e.g.:
+        with WACZClient(<path_to_wacz_file>) as wacz_client:
+            html_content = wacz_client.get_website_content_by_url("http://example.com")
+
+    This guarantees that any open ZipFile file objects will be closed upon context manager
+    exiting.
+    """
+
+    PAGES_FILEPATHS = ("pages/pages.jsonl", "pages/extraPages.jsonl")
+    CDX_INDEX_FILEPATH = "indexes/index.cdx.gz"
+    WARC_DIR = "archive"
+
+    def __init__(self, wacz_filepath: str):
+        self.wacz_filepath = wacz_filepath
+        self._via_context_manager = False
+        self._wacz_archive: zipfile.ZipFile | None = None
+        self._html_websites_df: pd.DataFrame | None = None
+
+    def __enter__(self) -> "WACZClient":
+        """Enter method for use when class instantiated via a context manager.
+
+        This sets the private attribute self._via_context_manager to True, allowing any
+        methods to know if operating with a context manager.
+        """
+        self._via_context_manager = True
+        return self
+
+    def __exit__(
+        self,
+        exc_type: type[BaseException] | None,
+        exc_value: BaseException | None,
+        traceback: TracebackType | None,
+    ) -> None:
+        """Exit method for use when class instantiated via a context manager.
+
+        When the context manager closes, this method closes any open file objects from the
+        WACZ ZipFile.
+        """
+        self.close()
+
+    def close(self) -> None:
+        """Close open WACZ ZipFile."""
+        if self._wacz_archive:
+            logger.debug("Closing WACZ ZipFile file handle.")
+            self._wacz_archive.close()
+            self._wacz_archive = None
+
+    @property
+    def wacz_archive(self) -> zipfile.ZipFile:
+        """Property that provides an open ZipFile file object of the target WACZ file.
+
+        NOTE: when accessed, this property will throw a ContextManagerRequiredError
+        exception if not operating in a context manager context.
+        """
+        if not self._via_context_manager:
+            msg = (
+                "Please use this class via a context manager for operations that read "
+                "the WACZ file"
+            )
+            raise ContextManagerRequiredError(msg)
+        if self._wacz_archive is None:
+            with smart_open.open(self.wacz_filepath, "rb") as file_obj:
+                _wacz_archive = zipfile.ZipFile(io.BytesIO(file_obj.read()))
+                self._wacz_archive = _wacz_archive
+        return self._wacz_archive
```

Comment:
While this obviously works, I'm not familiar with the private/public attribute switcheroo (forgive me for not knowing a better way to describe it) design pattern here. What's the advantage of the private attribute if it eventually becomes public?

Reply:
The advantage here, and also with the keyword extractor, is effectively caching an instance of that somewhat expensive object to create.  

The private `_wacz_archive` allows a place to store that object, so that if already loaded and saved (cached), it's available via the public, ergonomically friendly `self.wacz_archive`.

Reply:
Got it, thanks!